float32 farthest_distance
float32 closest_distance
